pkg/pillar/hypervisor/kvm.go

```diff
@@ -929,12 +939,180 @@ func getFmlCustomResolution(status *types.DomainStatus, globalConfig *types.Conf
 	return "", fmt.Errorf("invalid fml resolution %s", fmlResolutions)
 }
 
+type pciDevicesWithBridge struct {
+	bridgeBus string
+	devs      []*pciDevice
+}
+
+type multifunctionDevs map[string]*pciDevicesWithBridge
+
+func (md multifunctionDevs) isMultiFunction(p pciDevice) bool {
+	pciLongWOFunc, err := p.pciLongWOFunction()
+	if err != nil {
+		return false
+	}
+	devs, found := md[pciLongWOFunc]
+	return found && len(devs.devs) > 1
+}
+
+func (md multifunctionDevs) isFirstMultiFunction(p pciDevice) bool {
+	return md.index(p) == 0
+}
+
+func (md multifunctionDevs) index(p pciDevice) int {
+	pciLongWOFunc, err := p.pciLongWOFunction()
+	if err != nil {
+		logrus.Warnf("retrieving pci address without function failed: %v", err)
+		return -1
+	}
+	pciDevices, found := md[pciLongWOFunc]
+	if !found {
+		return -1
+	}
+
+	for i, dev := range pciDevices.devs {
+		if p.ioType == dev.ioType && p.pciLong == dev.pciLong {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func multifunctionDevGroup(pcis []pciDevice) multifunctionDevs {
+	mds := make(multifunctionDevs)
+
+	for i, pa := range pcis {
+		pciWithoutFunction, err := pa.pciLongWOFunction()
+		if err != nil {
+			logrus.Warnf("retrieving pci address without function failed: %v", err)
+			continue
+		}
+
+		_, found := mds[pciWithoutFunction]
+		if !found {
+			mds[pciWithoutFunction] = &pciDevicesWithBridge{
+				bridgeBus: "",
+				devs:      []*pciDevice{},
+			}
+		}
+		mds[pciWithoutFunction].devs = append(mds[pciWithoutFunction].devs, &pcis[i])
+	}
+
+	return mds
+}
+
+func (p pciDevice) pciLongWOFunction() (string, error) {
+	pciLongSplit := strings.Split(p.pciLong, ".")
+	if len(pciLongSplit) == 0 {
+		return "", fmt.Errorf("could not split %s", p.pciLong)
+	}
+	pciWithoutFunction := strings.Join(pciLongSplit[0:len(pciLongSplit)-1], ".")
+
+	return pciWithoutFunction, nil
+}
+
+type pciAssignmentsTemplateFiller struct {
+	multifunctionsDevices multifunctionDevs
+	file                  io.Writer
+}
+
+func (f *pciAssignmentsTemplateFiller) pciEBridge(pciID int, pciWOFunction string) error {
+	pciChildTemplateVars := struct {
+		PCIId int
+		Bus   int
+	}{
+		PCIId: 0,
+		Bus:   pciID,
+	}
+
+	tPCIeBridge, err := template.New("qemuPCIeBridge").Parse(qemuPCIPassthruBridgeTemplate)
+	if err != nil {
+		return fmt.Errorf("parsing qemuPCIPassthruBridgeTemplate failed: %w", err)
+	}
+	if err := tPCIeBridge.Execute(f.file, pciChildTemplateVars); err != nil {
+		return logError("can't write PCIe bridge Passthrough to config file (%v)", err)
```

Comment:
Do we need to return errors here if we do not handle them in the caller? And if we want to handle them - it would be nice to have one way to format the errors. Not two different options are used: `logError` here, and `fmt.Errorf` 3 lines above.

Reply:
I changed it to return an error and handle it in the caller.